core-backend/common/src/utils.rs

```diff
@@ -0,0 +1,117 @@
+use alloc::string::String;
+
+pub(crate) fn smart_truncate(s: &mut String, max_bytes: usize) {
+    let mut last_byte = max_bytes;
+
+    if s.len() > last_byte {
+        while !s.is_char_boundary(last_byte) && last_byte != 0 {
+            last_byte = last_byte.saturating_sub(1);
```

Comment:
We may omit the check against zero if we use `saturating_sub`. Or use just `-=`;

```rust
assert_eq!(0_usize.saturating_sub(1), 0);

Reply:
yeah, sure, but if we replace this check, we **theoretically** can run into infinity loop here: `while !s.is_char_boundary(last_byte)`, but this is unreachable, cause `assert!(any_string.is_char_boundary(0))`

this is really cheap check, so we can just left it as it is, isn't it?

Reply:
nvm, will remove it after master merge
no reason to keep excess checks, you're right